classes/local/datasource/images.php

```diff
@@ -217,6 +217,12 @@ public function get_instance_cache_key(): string {
         return '';
     }
 
+    /**
+     * Update the cache key fragment for the instance.
+     */
+    public function update_instance_cache_key() {
```

Comment:
I'm still not convinced this is the right way to do this. What I mean by 'this' is where we declare `update_instance_cache_key()` as an abstract method in the main `base` class but then most of the implementing classes just override with a method that does nothing essentially. If we need to make sure this method is implemented, can we please make sure it returns something at least? Or return void? Or update the phpdoc to make it clear this is overriding to do nothing?

Reply:
Do nothing is not necessarily a valid default. I have tired to encapsulate the different caching behaviours into traits to help make it easier to use and understand, but I do not consider any of them to constitute a 'default' behaviour. Thus the abstract methods. Even the ones that do nothing are specific circumstances, and do not constitute a default.

I have updated the docs to give more info, though.